### ISSUE TITLE
A memory leak when using bottleneck nanmax

### ISSUE THREAD

**Issue by aviolov** [pandas-dev/pandas#26485]
#### Code Sample, a copy-pastable example if possible

```python
import numpy as np
import pandas as pd
from bottleneck import nanmax
#from numpy import nanmax ~ using numpy.nanmax will not create a memory leak 
import gc
import tracemalloc

def pDbN_dosomething(df, rip_values = False):
    #ms = []    
    for col in df.columns:
        if rip_values:
            'If explicitly ripping the values of the column memory will NOT leak'
            m = abs(nanmax(df.loc[:, col].values))
        else:
            'If refering to the Series memory will leak'
            m = abs(nanmax(df.loc[:, col]))
            
        'random calculation to modify `df` - NOT required to trigger the memory leak'
        #df.loc[:, col] = df.loc[:, col] / df.iloc[:, 0]
        
        'persist `m` outside loop - NOT required to trigger the memory leak'    
        #ms.append(m)    
    return df

def printTraceStats():
    snapshot = tracemalloc.take_snapshot()
    top_stats = snapshot.statistics('lineno')
    print("[ Top 3 ]")
    for stat in top_stats[:3]:
        print(stat)
               
if __name__ == '__main__':   
    tracemalloc.start()
         
    D = pd.DataFrame(np.random.rand(1000000, 12))
    
    for _ in range(8):        
        df = D.copy()
        df = pDbN_dosomething(df, rip_values=True)        
        del df 
        gc.collect()    
    printTraceStats() 
    """
    [ Top 3 ]
    bug_PandasBottleneck_MemoryLeak.py:35: size=91.6 MiB, count=5, average=18.3 MiB  
    # this is initial allocation of the mother dataframe
    C:\tools\Anaconda3\envs\quant_env_r20190129_9d8c4b42\lib\_weakrefset.py:37: size=4224 B, count=32, average=132 B
    # not important
    ...
    """
    
    for _ in range(8):        
        df = D.copy()
        df = pDbN_dosomething(df, rip_values=False)        
        del df 
        gc.collect() 
    printTraceStats()
    r'''[ Top 3 ]
    C:\tools\Anaconda3\envs\quant_env_r20190129_9d8c4b42\lib\site-packages\pandas\core\internals.py:736: size=732 MiB, count=24, average=30.5 MiB
    # The df has now been duplicated 8 times 732 = 91*8
    C:\Users\alexandre.iolov\Documents\PySpace\NumPyBox\pandas_xarray\bug_PandasBottleneck_MemoryLeak.py:35: size=91.6 MiB, count=5, average=18.3 MiB
    # The original DataFrame from above
    C:\tools\Anaconda3\envs\quant_env_r20190129_9d8c4b42\lib\site-packages\pandas\core\internals.py:179: size=8960 B, count=176, average=51 B
    # not important 
    '''

```
#### Problem description
I am experiencing what appears to be a memory leak when using bottleneck.nanmax on a pandas Series. Calling `abs(bottleneck.nanmax(df.loc[:, col]))` will retain a reference to df in  memory even after `df` has been explicitly deleted and `gc.collect` called. This also sounds like a bottleneck issue, so let me know if I should cross-post. The issue is NOT present if I explicitly use `.values` on the Series or if I use `numpy.nanmax`

#### Expected Output
I expect the output, i.e. the total memory footprint when working on the `Series` to be the same as when working on the underlying `values`. 

#### Output of ``pd.show_versions()``

<details>

pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.5.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 79 Stepping 1, GenuineIntel
byteorder: little
LC_ALL: None
LANG: en
LOCALE: None.None

pandas: 0.22.0
pytest: 3.5.1
pip: 10.0.1
setuptools: 39.1.0
Cython: 0.28.2
numpy: 1.14.3
scipy: 1.1.0
pyarrow: None
xarray: 0.10.9
IPython: 6.4.0
sphinx: 1.7.4
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: 1.2.1
tables: 3.4.3
numexpr: 2.6.5
feather: None
matplotlib: 2.2.2
openpyxl: 2.5.3
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.4
lxml: 4.2.1
bs4: 4.6.0
html5lib: 1.0.1
sqlalchemy: 1.2.7
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

**Comment by TomAugspurger** [comment]
What do you observe with bottleneck.nanmax on an ndarray, either a raw ndarray created directly, or one extracted from a DataFrame with `df.loc[:, col]`. It'd be good to isolate between bottleneck / pandas' indexing / pandas' .max.

**Comment by aviolov** [comment]
I hoped the MRE would show that using bottleneck.nanmax on the df[].values does NOT result in a memory leak. That is my impression

 
 
  On Wed, 22 May 2019 at 21:57, Tom Augspurger<notifications@github.com> wrote:   
What do you observe with bottleneck.nanmax on an ndarray, either a raw ndarray created directly, or one extracted from a DataFrame with df.loc[:, col]. It'd be good to isolate between bottleneck / pandas' indexing / pandas' .max.

**Comment by VirosaLi** [comment]
Seem like it has been fixed right now for bottleneck 1.3.2. The output for the above example:

```bash
[ Top 3 ]
/home/xxx/PycharmProjects/test/test6.py:38: size=91.6 MiB, count=5, average=18.3 MiB
/home/xxx/miniconda3/envs/pandas-dev/lib/python3.8/abc.py:102: size=15.2 KiB, count=171, average=91 B
/home/xxx/PycharmProjects/test/test6.py:41: size=1552 B, count=2, average=776 B
[ Top 3 ]
/home/xxx/PycharmProjects/test/test6.py:38: size=91.6 MiB, count=5, average=18.3 MiB
/home/xxx/miniconda3/envs/pandas-dev/lib/python3.8/abc.py:102: size=15.2 KiB, count=171, average=91 B
/home/xxx/PycharmProjects/test/test6.py:17: size=1784 B, count=4, average=446 B
```

**Comment by jreback** [comment]
great
